crates/cxx-qt-lib/src/core/quuid.rs

```diff
@@ -0,0 +1,382 @@
+}
+
+impl fmt::Debug for QUuid {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        ffi::quuid_to_string(self).fmt(f)
+    }
+}
+
+impl QUuid {
+    /// Creates the null UUID. `to_string()` will output the null UUID as
+    /// "{00000000-0000-0000-0000-000000000000}".
+    pub const fn null() -> Self {
+        Self {
+            data1: 0,
+            data2: 0,
+            data3: 0,
+            data4: [0; 8],
+        }
+    }
+
+    /// This function returns a new UUID with variant `QUuidVariant::DCE` and version
+    /// `QUuidVersion::Md5`. `namespace` is the namespace and `data` is the basic data as described
+    /// by RFC 4122.
+    pub fn new_v3(namespace: &Self, data: &[u8]) -> Self {
+        ffi::quuid_new_v3(namespace, data)
+    }
+
+    /// On any platform other than Windows, this function returns a new UUID with variant
+    /// `QUuidVariant::DCE` and version `QUuidVersion::Random`. On Windows, a GUID is generated using
+    /// the Windows API and will be of the type that the API decides to create.
+    pub fn new_v4() -> Self {
+        ffi::quuid_new_v4()
+    }
+
+    /// This function returns a new UUID with variant `QUuidVariant::DCE` and version
+    /// `QUuidVersion::Sha1`. `namespace` is the namespace and `data` is the basic data as described
+    /// by RFC 4122.
+    pub fn new_v5(namespace: &Self, data: &[u8]) -> Self {
+        ffi::quuid_new_v5(namespace, data)
+    }
+
+    /// Creates a QUuid object from the string text, which must be formatted as five hex fields
+    /// separated by '-', e.g., "{xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx}" where each 'x' is a hex
+    /// digit. The curly braces shown here are optional, but it is normal to include them.
+    ///
+    /// If the conversion fails, a null UUID is returned.
+    pub fn from_string<'a, S>(s: S) -> Self
+    where
+        S: Into<crate::QAnyStringView<'a>>,
+    {
+        ffi::quuid_from_string(s.into())
+    }
+
+    /// Creates a UUID with the value specified by the parameters.
+    pub const fn from_fields(data1: u32, data2: u16, data3: u16, data4: [u8; 8]) -> Self {
+        Self {
+            data1,
+            data2,
+            data3,
+            data4,
+        }
+    }
+
+    pub const fn as_fields(&self) -> (u32, u16, u16, &[u8; 8]) {
+        (self.data1, self.data2, self.data3, &self.data4)
+    }
+
+    const fn to_big_endian(self) -> Self {
+        #[cfg(target_endian = "big")]
+        {
+            self
+        }
+        #[cfg(target_endian = "little")]
+        {
+            Self {
+                data1: self.data1.swap_bytes(),
+                data2: self.data2.swap_bytes(),
+                data3: self.data3.swap_bytes(),
+                data4: self.data4,
+            }
+        }
+    }
+
+    /// Creates a UUID from its representation as a byte array in big endian.
+    pub const fn from_bytes(bytes: [u8; 16]) -> Self {
+        unsafe { mem::transmute::<[u8; 16], Self>(bytes) }.to_big_endian()
```

Comment:
I have to admit it took me quite a while to understand this.

The function takes an array of big-endian bytes and then calls `to_big_endian` on it. That is quite confusing, as it's already big-endian.
I believe this still turns out to be correct, as the internal byte order is actually the host endianness, so calling to_big_endian will actually go from big to little endian, but only if the host order is little endian, which is exactly what's needed. But it's definitely not obvious that this is whats happening.

So at the very least this needs an explanation comment.

Reply:
I've added explaining documentation and directly copied over the naming convention and documentation for [u8::to_be](https://doc.rust-lang.org/std/primitive.u8.html#method.to_be), so that it will be more familiar to people who have worked with endian conversions. Do these changes make it sufficiently easy to understand? Is there anything else I should add?

Reply:
Yes, thank you.